app/controllers/authenticate_controller.rb

```diff
@@ -60,23 +48,38 @@ def k8s_inject_client_cert
 
   private
 
+  def authentication_strategy
+    @authentication_strategy ||= ::Authentication::Strategy.new(
+      authenticators: installed_authenticators,
+      audit_log: ::Authentication::AuditLog,
+      security: nil,
+      env: ENV,
+      role_cls: ::Role,
+      token_factory: TokenFactory.new
+    )
+  end
+
+  def authentication_input
+    ::Authentication::Strategy::Input.new(
+      authenticator_name: params[:authenticator],
+      service_id:         params[:service_id],
+      account:            params[:account],
+      username:           params[:id],
+      password:           request.body.read,
+      origin:             request.ip,
```

Comment:
Occurs to me now the `origin` is a bit vague, especially when you see it in a context like

> "User is not authorized to login from the current origin"

Does that mean a particular IP address?  A country?  A network?  Looks like you just lifted the name from existing code on the `Role` class, but I'd say no reason to let a bad name spread.  I'd suggest `ip_address` or `request_ip` instead...